[torch/elastic] Introduce `DynamicRendezvousHandler` and `RendezvousBackend`.

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55637 [torch/elastic] Introduce `EtcdRendezvousBackend`.
* #55636 [torch/elastic] Introduce `C10dRendezvousBackend`.
* **#55635 [torch/elastic] Introduce `DynamicRendezvousHandler` and `RendezvousBackend`.**

This diff introduces the `DynamicRendezvousHandler` type as a stub implementation and its accompanying `RendezvousBackend` interface.

`DynamicRendezvousHandler` is intended to be a backend-agnostic type that will contain the core (bulk) logic of rendezvous handling. Any backend specific operation will be delegated to a concrete subclass of `RendezvousBackend` (e.g. `C10dRendezvousBackend` - see D27654492) that is passed as a constructor argument to `DynamicRendezvousHandler`.

Differential Revision: [D27654478](https://our.internmc.facebook.com/intern/diff/D27654478/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27654478/)!

cbalioglu committed:
[torch/elastic] Introduce `DefaultRendezvousHandler` and `RendezvousBackend`.

This diff introduces the `DefaultRendezvousHandler` type as a stub implementation and its accompanying `RendezvousBackend` interface.

`DefaultRendezvousHandler` is intended to be a backend-agnostic type that will contain the core (bulk) logic of rendezvous handling. Any backend specific operation will be delegated to a concrete subclass of `RendezvousBackend` (e.g. `C10dRendezvousBackend` - see D27654492) that is passed as a constructor argument to `DefaultRendezvousHandler`.

Differential Revision: [D27654478](https://our.internmc.facebook.com/intern/diff/D27654478/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27654478/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/elastic/rendezvous/api.py:
```diff
@@ -43,93 +42,80 @@ class RendezvousHandler(abc.ABC):
-        which means all future attempts to re-rendezvous (within same job) will
-        fail.
-
-        .. note:: ``is_closed`` and ``set_closed`` have semantics of eventual
```
Is there a reason this note was removed?

cbalioglu replied:
The note is still there actually. I just removed the "note::" marking and included the description in the main description.

pritamdamania87 commented on torch/distributed/elastic/rendezvous/default_rendezvous.py:
```diff
@@ -0,0 +1,293 @@
+        """See base class."""
+        return self._backend.name
+
+    def next_rendezvous(self) -> Tuple[Store, int, int]:
+        """See base class."""
+        raise NotImplementedError()
+
+    def is_closed(self) -> bool:
+        """See base class."""
+        raise NotImplementedError()
+
+    def set_closed(self) -> None:
+        """See base class."""
+        raise NotImplementedError()
+
+    def num_nodes_waiting(self) -> int:
+        """See base class."""
+        raise NotImplementedError()
+
+    def get_run_id(self) -> str:
+        """See base class."""
+        return self._run_id
+
+    def shutdown(self) -> bool:
+        """See base class."""
+        raise NotImplementedError()
```
I might be missing something here, but the PR says the bulk of the rendezvous logic would be in this class. But its not implemented here and I don't see it in the other two PRs on top of it (C10dRendezvous and EtcRendezvous backends). Is there another PR in the works that adds this functionality?

cbalioglu replied:
Yes, I briefly mentioned it in the description as "as a stub implementation". It is actually the core piece of this whole task and will have its own PR by tomorrow.

H-Huang commented on test/distributed/elastic/rendezvous/default_rendezvous_test.py:
```diff
@@ -0,0 +1,177 @@
+        join_timeouts = [timedelta(seconds=0), timedelta(seconds=-1)]
+
+        for join_timeout in join_timeouts:
+            with self.subTest(join_timeout=join_timeout):
```
neat! I haven't seen `subTest` before

H-Huang reviewed (approved):
Left some comments on internal diff, accepting here. 

Just to confirm, the newly added tests for torch/elastic have yet to be included as part of OSS CI right? They are only run internally?

cbalioglu committed:
Update on "[torch/elastic] Introduce `DefaultRendezvousHandler` and `RendezvousBackend`."


This diff introduces the `DefaultRendezvousHandler` type as a stub implementation and its accompanying `RendezvousBackend` interface.

`DefaultRendezvousHandler` is intended to be a backend-agnostic type that will contain the core (bulk) logic of rendezvous handling. Any backend specific operation will be delegated to a concrete subclass of `RendezvousBackend` (e.g. `C10dRendezvousBackend` - see D27654492) that is passed as a constructor argument to `DefaultRendezvousHandler`.

Differential Revision: [D27654478](https://our.internmc.facebook.com/intern/diff/D27654478/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27654478/)!

[ghstack-poisoned]

cbalioglu committed:
Update on "[torch/elastic] Introduce `DefaultRendezvousHandler` and `RendezvousBackend`."


This diff introduces the `DefaultRendezvousHandler` type as a stub implementation and its accompanying `RendezvousBackend` interface.

`DefaultRendezvousHandler` is intended to be a backend-agnostic type that will contain the core (bulk) logic of rendezvous handling. Any backend specific operation will be delegated to a concrete subclass of `RendezvousBackend` (e.g. `C10dRendezvousBackend` - see D27654492) that is passed as a constructor argument to `DefaultRendezvousHandler`.

Differential Revision: [D27654478](https://our.internmc.facebook.com/intern/diff/D27654478/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27654478/)!

[ghstack-poisoned]

tierex commented on torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:
```diff
@@ -0,0 +1,299 @@
+    +-------------------+------------------------------------------------------+
+    | Parameter         | Description                                          |
+    +===================+======================================================+
+    | join_timeout      | The total time, in seconds, within which the         |
```
Do you think maybe it makes sense to use a single timeout for join_timeout, last_call_timeout and close_timeout ?
I wonder when users would need to explicitly set each one separately.

cbalioglu replied:
I think in most of the cases (if at all) the user would be just adjusting `join_timeout` only. Combining these timeouts into one does not make much sense in my opinion as they are used for different purposes. `close_timeout` is used for `set_closed()` and `shutdown()` and should ideally be much shorter than `join_timeout`. `last_call_timeout` on the other hand is used for early completion of the rendezvous and is mostly independent of `join_timeout`. `join_timeout` more or less determines how long a call to `next_rendezvous()` should take.
